File: src/02_algorithms/02_searching/02_binary_search/implementations/ts/02_binary_search.ts
export const binarySearch = (
  items: Array<number>,
  target: number,
  onStep?: (element) => void
): number => {
  let leftIndex = 0;
  let rightIndex = items.length - 1;

  while (leftIndex <= rightIndex) {
    const middleIndex = Math.floor((leftIndex + rightIndex) / 2);
    onStep(items[middleIndex]);

    if (items[middleIndex] === target) return target;
    if (items[middleIndex] < target) leftIndex = middleIndex + 1;
    if (items[middleIndex] > target) rightIndex = middleIndex - 1;
  }

  return null;
};

File: src/02_algorithms/02_searching/02_binary_search/implementations/go/02_binary_search.go
package binarySearch

import "fmt"

func BinarySearch(items []int, target int) (int, error) {
	leftIndex := 0
	rightIndex := len(items) - 1

	for leftIndex <= rightIndex {
		middleIndex := (leftIndex + rightIndex) / 2

		if target == items[middleIndex] {
			return target, nil
		} else if target < items[middleIndex] {
			rightIndex = middleIndex - 1
		} else {
			leftIndex = middleIndex + 1
		}
	}

	return -1, fmt.Errorf("index out of bounds")
}
